pkg/sql/txn_instrumentation.go

```diff
@@ -105,17 +123,31 @@ type txnInstrumentationHelper struct {
 }
 
 func (h *txnInstrumentationHelper) StartDiagnostics(
-	txnRequest stmtdiagnostics.TxnRequest, reqID stmtdiagnostics.RequestID,
+	txnRequest stmtdiagnostics.TxnRequest, reqID stmtdiagnostics.RequestID, span *tracing.Span,
 ) {
-	h.diagnosticsCollector = newTxnDiagnosticsCollector(txnRequest, reqID)
+	h.diagnosticsCollector = newTxnDiagnosticsCollector(txnRequest, reqID, span)
 }
 
+// DiagnosticsInProgress returns true if diagnostics collection is considered
+// to be progress. This is determined by the following:
+// - there is an active request id
+// - there has been at least 1 statement bundle collected
+// - there are still statements to be collected
+// If none of these are true, then diagnostics collection is not in progress.
+//
+// NB: Currently, transaction diagnostics collection only happens through
+// requests, so the first condition is sufficient. If, in the future, new
+// methods of collecting transaction diagnostics are added, such as through
+// "explain analyze (debug)", then the other two conditions will be necessary.
 func (h *txnInstrumentationHelper) DiagnosticsInProgress() bool {
-	return h.diagnosticsCollector.requestId != 0
+	return h.diagnosticsCollector.requestId != 0 ||
+		len(h.diagnosticsCollector.stmtBundles) > 0 ||
+		len(h.diagnosticsCollector.stmtsFpsToCapture) > 0
 }
 
 func (h *txnInstrumentationHelper) Reset() {
 	if h.diagnosticsCollector.requestId != 0 {
+		h.diagnosticsCollector.span.SetRecordingType(tracingpb.RecordingOff)
```

Comment:
If we're Resetting, shouldn't the span be set to nil so we can get a new one?

Reply:
@dhartunian we are turning off recording for the span so that nothing is recorded for the remainder of the transaction since we are no longer collecting a transaction diagnostic. When we reset, the diagnostic is reset to the zero value to tell the instrumentation helper that we will no longer collect diagnostics for this transaction and the request is "released" to let a new transaction attempt to fulfill it. We will never need to get a new span because the transaction diagnostics has been aborted. 

when a new instrumentationHelper is created for a new transaction, it will start a new span for that transaction and attach it to the diagnosticsCollector created for that execution.